### PR TITLE
Limit AXFR to whitelisted nodes

### LINKED ISSUE #2647: DNS: Implement whitelisting for AXFR
We should have a way to restrict AXFR to whitelisted clients, so that it doesn't get abused.
A config file entry, probably at the registry level, as it doesn't make much sense to have it at the zone level, would be needed.

```yaml
registry:
  enabled: true
  whitelist:
    - 1.2.3.4
    - foo.bar
```

### PR DESCRIPTION
**Pull request by mkykadir**
Fixes #2647

### CODE REVIEW

File: source/agora/node/Runner.d [hunk 1]
```diff
@@ -319,7 +319,7 @@ private void runDNSServer_canThrow (in RegistryConfig config, NameRegistry regis
             auto pack = udp.recv(buffer, ppeer);
             auto query = deserializeFull!Message(pack);
             registry.answerQuestions(
-                query,
+                query, peer.toAddressString,
```
Comment (mkykadir) [inline source/agora/node/Runner.d:322]: Need to consider [this](https://github.com/bosagora/agora/pull/2703#discussion_r780897531)

Comment (mkykadir) [reply source/agora/node/Runner.d:322]: I will also consider reverse proxy situation

File: source/agora/node/Runner.d [hunk 2]
```diff
@@ -420,7 +420,7 @@ private void runTCPDNSServer_canThrow (TCPConnection conn, NameRegistry registry
 
         auto query = deserializeFull!Message(buffer[0 .. size]);
         registry.answerQuestions(
-            query,
+            query, conn.peerAddress(),
```
Comment (mkykadir) [inline source/agora/node/Runner.d]: this returns with port number also

File: source/agora/node/Runner.d (hunk 1 above)
Comment (mkykadir) [reply source/agora/node/Runner.d:322]: I've checked Nginx about reverse proxy on stream sockets (i.e. TCP and UDP) and stream configs doesn't have a `forwarded for` feature like in the HTTP configurations. Either client IP is directed or proxy IP is directed, thus we cannot do anything about it through Agora.

**Comment by mkykadir** [comment]
Rebased

File: source/agora/node/Registry.d [hunk 3]
```diff
@@ -56,8 +56,16 @@ public class NameRegistry: NameRegistryAPI
...
+    private enum ZoneIndex
+    {
+        REALM = 0,
+        VALIDATOR = 1,
+        FLASH = 2,
+    }
+
     /// Zones of the registry
-    private ZoneData[Domain] zones;
+    private ZoneData[3] zones;
```
Comment (omerfirmak) [inline source/agora/node/Registry.d]: Would `ZoneIndex.max + 1` work?

**Review by linked0 — APPROVED** [review]
It's green. LGTM.

File: source/agora/node/Registry.d [hunk 4]
```diff
@@ -56,8 +56,16 @@ public class NameRegistry: NameRegistryAPI
     ///
     protected RegistryConfig config;
 
+    /// Indexes for `zones`
+    private enum ZoneIndex
+    {
+        REALM = 0,
```
Comment (Geod24) [inline source/agora/node/Registry.d]: Why the uppercase ?

File: source/agora/node/Registry.d [hunk 5]
```diff
@@ -438,9 +445,10 @@ public class NameRegistry: NameRegistryAPI
 
     auto findZone (Domain name, bool matches = true) @safe
     {
-        auto matching_zone = name in this.zones;
-        if (matching_zone)
-            return matches ? &matching_zone.answer_matches : &matching_zone.answer_owns;
+        foreach (i, zone; this.zones)
```
Comment (Geod24) [inline source/agora/node/Registry.d]: `const ref zone` to avoid copying ?

File: source/agora/node/Config.d [hunk 6]
```diff
@@ -514,6 +514,9 @@ public struct ZoneConfig
     /// CNAME of the primary server for this zone
     public SetInfo!string primary;
 
+    /// Servers that are allowed to do AXFR queries, applicable for authoritatives
+    public @Optional immutable string[] allow_transfer;
```
Comment (Geod24) [inline source/agora/node/Config.d]: Need to validate that it's using IP addresses tho

**Review by Geod24 — APPROVED** [review]
Broadly LGTM.
Please add validation on the config level so it's hard / impossible to misconfigure.

File: source/agora/node/Config.d [hunk 7]
```diff
@@ -467,6 +469,16 @@ public struct RegistryConfig
...
 
+            foreach (addr; zone.allow_transfer)
+            {
+                try
+                    parseAddress(addr);
+                catch (Exception)
+                    ensure(0,
+                        "allow_transfer: Field must consist from IP addresses, {} is not an IP address",
+                        addr);
+            }
```
Comment (Geod24) [inline source/agora/node/Config.d]: I think you could get away with `guessAddressType`, which doesn't `throw`.
Also if you create a simple type for it with a `fromString` / `this (string)` and throw from there, you will get the file/line and field name printed for you by the config parser.

Comment (mkykadir) [reply source/agora/node/Config.d]: I picked second approach, which looks better during user configuration I think.

**Review by Geod24 — APPROVED** [review]

**Pull request merged**